framework/src/media/StreamHandler.cpp

```diff
@@ -0,0 +1,190 @@
+#endif
+
+#ifndef CONFIG_HANDLER_STREAM_BUFFER_THRESHOLD
+#define CONFIG_HANDLER_STREAM_BUFFER_THRESHOLD 2048
+#endif
+
+namespace media {
+namespace stream {
+
+StreamHandler::StreamHandler() :
+	mWorker(0),
+	mWorkerName("StreamWorker"),
+	mWorkerStackSize(4096),
+	mIsWorkerAlive(false)
+{
+}
+
+bool StreamHandler::open()
+{
+	if (!getStreamBuffer()) {
+		auto streamBuffer = StreamBuffer::Builder()
+								.setBufferSize(CONFIG_HANDLER_STREAM_BUFFER_SIZE)
+								.setThreshold(CONFIG_HANDLER_STREAM_BUFFER_THRESHOLD)
+								.build();
+
+		if (!streamBuffer) {
+			meddbg("streamBuffer is nullptr!\n");
+			return false;
+		}
+
+		setStreamBuffer(streamBuffer);
+	}
+
+	if (!(mDataSource->open() && registerCodec(mDataSource->getAudioType(), mDataSource->getChannels(), mDataSource->getSampleRate()))) {
+		return false;
+	}
+
+	return start();
+}
+
+bool StreamHandler::close()
+{
+	stop();
+	unregisterCodec();
+	return mDataSource->close();
+}
+
+bool StreamHandler::start()
+{
+	medvdbg("StreamHandler::start()\n");
+	if (mDataSource && !mDataSource->isPrepared()) {
+		return false;
+	}
+
+	createWorker();
+	return mIsWorkerAlive;
+}
+
+bool StreamHandler::stop()
+{
+	destroyWorker();
+	return true;
+}
+
+void StreamHandler::createWorker()
+{
+	medvdbg("StreamHandler::createWorker()\n");
+	if (mStreamBuffer && !mIsWorkerAlive) {
+		resetWorker();
+		mStreamBuffer = nullptr;
+		mIsWorkerAlive = true;
+
+		pthread_attr_t attr;
+		pthread_attr_init(&attr);
+		pthread_attr_setstacksize(&attr, mWorkerStackSize);
+		int ret = pthread_create(&mWorker, &attr, static_cast<pthread_startroutine_t>(StreamHandler::workerMain), this);
+		if (ret != OK) {
+			meddbg("Fail to create StreamHandler Worker thread, return value : %d\n", ret);
+			mIsWorkerAlive = false;
+			return;
+		}
+		pthread_setname_np(mWorker, mWorkerName);
+	}
+}
+
+void StreamHandler::destroyWorker()
+{
+	medvdbg("StreamHandler::destoryWorker()\n");
+	if (mIsWorkerAlive) {
+		// Setup flag,
+		mIsWorkerAlive = false;
+
+		// Worker may be blocked in buffer reading.
+		mBufferWriter->setEndOfStream();
+
+		// Wake worker up,
+		wakenWorker();
+
+		// Join thread.
+		pthread_join(mWorker, NULL);
+	}
+}
+
+void StreamHandler::sleepWorker()
+{
+	std::unique_lock<std::mutex> lock(mMutex);
+	// In case of overrun, DO NOT sleep worker.
+	if (mIsWorkerAlive) {
+		mCondv.wait(lock);
+	}
+}
+
+void StreamHandler::wakenWorker()
+{
+	std::lock_guard<std::mutex> lock(mMutex);
+	mCondv.notify_one();
+}
+
+void *StreamHandler::workerMain(void *arg)
+{
+	medvdbg("StreamHandler::workerMain()\n");
+
+	if (arg == nullptr) {
+		meddbg("%s[line : %d] Fail : arg is nullptr\n", __func__, __LINE__);
+		return NULL;
+	}
+
+	auto stream = static_cast<StreamHandler *>(arg);
+
+	while (stream->mIsWorkerAlive) {
+		// Waken up by a writing/flushing/stopping operation
+		stream->sleepWorker();
+
+		// Worker may be stoped
+		if (!stream->mIsWorkerAlive) {
+			break;
+		}
+
+		if (!stream->processWorker()) {
```

Comment:
If some of if state is very simple then you'd better merge it to decrease code size.
But if any of state is very important, then put some log & keep it separately.

Reply:
The logic of processWorker depends on derived classes.
So it seems hard to merge it.

Reply:
I mean below
```
if ((!stream->mIsWorkerAlive) || (!stream->processWorker())) {
   break;
}
```
or
```
if (!stream->mIsWorkerAlive) {
   medvdbg("Worker is dead, break\n");
   break;
}

if (!stream->processWorker()) {
    mevdbg("there's no process in worker\n");
   break;
}

```

Reply:
Oh... I got it. Fixed it